tsl/src/remote/copy_fetcher.c

```diff
@@ -338,85 +385,101 @@ copy_fetcher_complete(CopyFetcher *fetcher)
-					 * when reading a NULL input value. The function must still
-					 * return NULL in this case, unless it raises an error.
-					 * (This case is mainly meant to support domain receive
-					 * functions, which might need to reject NULL inputs.)
-					 * https://www.postgresql.org/docs/current/sql-createtype.html
-					 */
-					if (!attconv->conv_funcs[att].fn_strict)
-					{
-						values[att] = ReceiveFunctionCall(&attconv->conv_funcs[att],
-														  NULL,
-														  attconv->ioparams[att],
-														  attconv->typmods[att]);
-					}
-					else
+					nulls[i] = true;
+				}
+
+				MemoryContextSwitchTo(fetcher->state.tuple_mctx);
+				for (int i = 0; i < retrieved_natts; i++)
+				{
+					const int att = list_nth_int(retrieved_attrs, i) - 1;
+					Assert(att >= 0);
+					Assert(att < tupdesc_natts);
+					const int32 att_bytes = copy_data_read_int32(&copy_data);
+					if (att_bytes == -1)
 					{
-						values[att] = PointerGetDatum(NULL);
+						/*
+						 * NULL. From the Postgres docs:
+						 * Usually, a receive function should be declared STRICT; if
+						 * it is not, it will be called with a NULL first parameter
+						 * when reading a NULL input value. The function must still
+						 * return NULL in this case, unless it raises an error.
+						 * (This case is mainly meant to support domain receive
+						 * functions, which might need to reject NULL inputs.)
+						 * https://www.postgresql.org/docs/current/sql-createtype.html
+						 */
+						if (!attconv->conv_funcs[att].fn_strict)
+						{
+							values[att] = ReceiveFunctionCall(&attconv->conv_funcs[att],
+															  NULL,
+															  attconv->ioparams[att],
+															  attconv->typmods[att]);
+						}
+						else
+						{
+							values[att] = PointerGetDatum(NULL);
+						}
+						nulls[att] = true;
+						continue;
 					}
-					nulls[att] = true;
-					continue;
-				}
 
-				StringInfoData att_data = { 0 };
-				att_data.data = copy_data_read_bytes(&copy_data, att_bytes);
-				att_data.len = att_bytes;
+					StringInfoData att_data = { 0 };
+					att_data.data = copy_data_read_bytes(&copy_data, att_bytes);
+					att_data.len = att_bytes;
 
-				values[att] = ReceiveFunctionCall(&attconv->conv_funcs[att],
-												  &att_data,
-												  attconv->ioparams[att],
-												  attconv->typmods[att]);
-				nulls[att] = false;
+					values[att] = ReceiveFunctionCall(&attconv->conv_funcs[att],
+													  &att_data,
+													  attconv->ioparams[att],
+													  attconv->typmods[att]);
+					nulls[att] = false;
+				}
 			}
-
 			MemoryContextSwitchTo(fetcher->state.batch_mctx);
-
 			PQfreemem(copy_data.data);
+			dataptr = NULL;
 		}
 
-		fetcher->state.num_tuples = row;
+		/* Don't count the file trailer as a row if this was the last batch */
+		fetcher->state.num_tuples = fetcher->file_trailer_received ? row - 1 : row;
 		fetcher->state.next_tuple_idx = 0;
+
```

Comment:
Should we check that the file trailer is received here as well?

Reply:
Github is being weird, I mean in the if below

Reply:
You mean to assert it?

Reply:
yep

Reply:
Actually, EOF can be read without the file trailer being received. For example, when an error or cancel happens at the other end.

Reply:
Looks like we get to this branch only when there's no error. EOF with error does `ereport` earlier.